[KIP-848] User documentation (#4702)
confluentinc/librdkafka: INTRODUCTION.md

Reviewer: I think we need to adjust few sections and move some part to the changelog instead of here.

To read more about static group membership, see [KIP-345](https://cwiki.apache.org/confluence/display/KAFKA/KIP-345%3A+Introduce+static+membership+protocol+to+reduce+consumer+rebalances).

### Next generation of the consumer group protocol: [KIP 848](https://cwiki.apache.org/confluence/display/KAFKA/KIP-848%3A+The+Next+Generation+of+the+Consumer+Rebalance+Protocol)

> Starting from librdkafka 2.4.0 the next generation consumer group rebalance protocol
> is in **Early Access**. It means it's still  _not production-ready_ and
> _not supported_, given it's still under validation and lacking some needed features.

Reviewer: ```suggestion
Starting from librdkafka 2.4.0 the next generation consumer group rebalance protocol introduced in KIP-848
is introduced. 

Warning: It is still in **Early Access** which means it's still  _not production-ready_ and
_not supported_, given it's still under validation and lacking some needed features. Some 
of the feature and contracts might change in future.
```

With this protocol the role of the Group Leader (a member) is removed and
the assignment is calculated by the Group Coordinator (a broker) and sent
to each member through heartbeats.

To test it, a Kafka cluster must be set up, in KRaft mode, and the new group
protocol enabled with the `group.coordinator.rebalance.protocols` property.

> Broker version must be Apache Kafka 3.7.0 or newer.

Reviewer: Let's add a reference to the broker documentation for it. https://cwiki.apache.org/confluence/display/KAFKA/The+Next+Generation+of+the+Consumer+Rebalance+Protocol+%28KIP-848%29+-+Early+Access+Release+Notes

Client side, it can be enabled by setting the new property `group.protocol=consumer`.
A second property named `group.remote.assignor` is added to choose desired
remote assignor.

**Available features**

- Subscription to one or more topics
- Rebalance callbacks (see contract changes)
- Static group membership
- Configure remote assignor
- Max poll interval is enforced
- Offline upgrade from an empty consumer group with committed offsets

**Missing features**

- Regular expression support when subscribing
- AdminClient changes as described in the KIP
- Client side assignors as described in the KIP

> - Online upgrade of a non-empty consumer group

Reviewer: I don't know if we should call these two as missing features or not. Adding them to missing features shows that we are planning to implement these in future which we are not as of now.

Instead of "Missing Features" we should call them "Future features" and remove these two points from there.

**Contract changes**

Along with the new feature there are some needed contract changes,
so the protocol will be enabled by default only with a librdkafka major release.

>  - These client properties are deprecated: `partition.assignment.strategy`,
>  replaced by `group.remote.assignor`;
> `session.timeout.ms` and `heartbeat.interval.ms`, replaced by broker configurations
> `group.consumer.session.timeout.ms` and `group.consumer.heartbeat.interval.ms`.

Reviewer: ```suggestion
 - Deprecated client configurations for the new protocol
     - `partition.assignment.strategy` replaced by `group.remote.assignor`
     - `session.timeout.ms` replaced by broker configuration `group.consumer.session.timeout.ms`
     - `heartbeat.interval.ms`, replaced by broker configuration `group.consumer.heartbeat.interval.ms`
     - `group.protocol.type` which is not used in the new protocol
```

>  - Protocol rebalance is fully incremental, so the only allowed functions to
>    use in a rebalance callback will be `rd_kafka_incremental_assign` and
>    `rd_kafka_incremental_unassign`. Currently you can still use existing code
>    and the expected function to call is determined based on the chosen
>    `partition.assignment.strategy` but this will be removed already in next
>    release.

Reviewer: ```suggestion
 - Protocol rebalance is fully incremental, so the only allowed functions to
   use in a rebalance callback will be `rd_kafka_incremental_assign` and
   `rd_kafka_incremental_unassign`. Currently you can still use existing code
   and the expected function to call is determined based on the chosen
   `partition.assignment.strategy` but this will be removed in the next
   release.
```

   When setting the `group.remote.assignor` property, it's already

>    required to use the incremental assign and unassign functions.

Reviewer: We should also explain that even `range` assignor is sticky in the new protocol.

 - With a static group membership, if two members are using the same

>    `group.instance.id`, the one that joins the consumer group last will be

Reviewer: ```suggestion
   `group.instance.id`, the one that joins the consumer group later will be
```

   fenced, with the fatal `UNRELEASED_INSTANCE_ID` error. Before, it was the existing
   member to be fenced. This was changed to avoid two members contending the
   same id. It also means that any instance that crashes won't be automatically
   replaced by a new instance until session times out and it's especially required
   to check that consumers are being closed properly on shutdown. Ensuring that
   no two instances with same `group.instance.id` are running at any time
   is also important.

 - Session timeout is remote only and, if the Coordinator isn't reachable
   by a member, this will continue to fetch messages, even if it won't be able to
   commit them. Otherwise, the member will be fenced as soon as it receives an
   heartbeat response from the Coordinator.
   With `classic` protocol, instead, member stops fetching when session timeout
   expires on the client.

   For the same reason, when closing or unsubscribing with auto-commit set,
   the member will try to commit until a specific timeout has passed.
   Currently the timeout is the same as the `classic` protocol and it corresponds
   to the deprecated `session.timeout.ms`, but it will change before the feature

>    reaches a stable state.

Reviewer: I don't think we should explain that we are still using deprecated property.

 - An `UNKNOWN_TOPIC_OR_PART` error won't be received when a consumer is
   subscribing to a topic that doesn't exist in local cache, as the consumer

>    is still subscribing to the topic and it could be created just after that.

Reviewer: We should explain more on why we are writing this. What used to happen in old protocol.

 - A consumer won't do a preliminary Metadata call that returns a
   `TOPIC_AUTHORIZATION_FAILED`, topic partitions will be assigned to the member

>    by the Coordinator only if it's authorized to consume from the topic.

Reviewer: Same

 - Number of assign/revoke callbacks isn't fixed anymore, as it depends on

>    heartbeat timing.

Reviewer: What does this mean and does it affect the users?
